### PR TITLE
Add order items directly to Stripe subscription

### LINKED ISSUE #2893: Use Stripe `Product`s for misc subscription costs such as shipping or fees, instead of ad-hoc invoice items
### Description
Right now, we are creating ad-hoc invoice items prior to each renewal in Stripe. We were taking this approach because we believed there was too much overhead with adding these items directly to the subscription. For example, we thought we would need to maintain a separate product and price object for each different shipping, fee, etc. in WC. 

After some discussion with Stripe, it turns out there is a workaround to avoid all of this overhead. Mainly, creating a sort of generic product for each of the item types and adding them to the subscription with ad-hoc price data.

### Acceptance criteria

* Invoice items should be added directly to subscriptions
* Subscription items should remain flexible to account for changes in WC
* Subscription behavior and functionality should remain unchanged

### Additional context
See pdi1mn-11-p2 for meeting notes.

### PR DESCRIPTION
**Pull request by chickenn00dle**
Fixes #2893 

#### Changes proposed in this Pull Request

- Add shipping and fee items to a subscription at creation
- Add Stripe coupons to a subscription for WC discounts (This is pending our WCPay accounts being gated into the multiple coupons feature by Stripe)
- Add validation at `invoice.upcoming` webhook event

This changes how subscriptions items are added. These were previously added as invoice items during one of the renewal webhook events. But this PR removes that functionality and instead adds items directly to a subscription when it is created in Stripe.

With this approach, we also have to create coupons for WC discounts since the subscriptions endpoint does not allow for negative item amounts.

Note that going with this approach means we need to add a generic product in Stripe for each of our WC order item types, which we store in `wp_options`.

#### Testing instructions

This PR relies on this WCPay Server PR: 1172-gh-Automattic/woocommerce-payments-server

- Make sure the WCPay Server PR mentioned above is checked out and local server is running
- Make sure you have a subscription product that has been synced to Stripe
- Make sure this product has some shipping cost and or sign-up fee associated with it
- Check out the product and verify the item data in WC and Stripe match.
- Repeat the above for various subscriptions with different settings (free trial, synchronized, etc.)

To test validation, you will need to use Billing Clocks via the dev tools plugin.

- Enable billing clocks on a subscription in WP-Admin (WooCommerce > Subscriptions > your subscription)
- DON'T TRIGGER THE UPCOMING INVOICE YET. Go to your Stripe dashboard and update the quantity and tax rates for all items on the matching subscription: ([#](https://cloudup.com/cBd3CLUjoxf), [#](https://cloudup.com/cqg6JCWd0AX))
- Now trigger the upcoming invoice and confirm the correct values have been set again in Stripe.

-------------------

- [ ] Added changelog entry (or does not apply)
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

- [ ] Added testing instructions to the [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) (or does not apply)

### CODE REVIEW

**Comment by chickenn00dle** [comment]
This PR is all but ready. Before we can test coupons though, we need to wait for `multiple coupon` functionality to be added to our accounts: pdi1mn-8-p2#comment-30

I'm gonna go ahead and label this ready for review though because it's probably a good idea to have some eyes on this while we wait.

Basically everything but adding coupons to a subscription (as well as validating coupon data on `invoice.upcoming` is ready to go, so if anyone does start looking at this, it's expected for these cases to fail. The testing instructions above are updated to reflect this.

**Comment by james-allan** [comment]
Nice work on this @chickenn00dle! 

I've tested a few scenarios and everything is looking pretty good. I've just got a couple of questions and an issue below:

**Subscriptions with multiple fees:**

1. Add the code snippet below. 
2. Purchase a subscription. (https://d.pr/i/Q1MGcp)
3. View the subscription in the Stripe Dashboard. (https://d.pr/i/Q1MGcp)

Question: Is there a way that we can add something to make this fee distinguishable? I had a look in the Stripe API and I can't see anything to suggest we can add price data meta. So maybe we can't. 

_This code snippet adds fees to the recurring cart. No fees are charged on the initial order._
```php
add_filter( 'woocommerce_cart_calculate_fees', 'add_recurring_postage_fees', 10, 1 );

function add_recurring_postage_fees( $cart ) {
    if ( ! empty( $cart->recurring_cart_key ) ) {
        $cart->add_fee( 'Postage', 5 );
	$cart->add_fee( 'Handling', 15 );
    }
}
```

 **Subscriptions with multiple fees:**

1. Add the code snippet above.
2. Purchase a subscription. 
3. View the Subscription's first invoice in Stripe. 
     - Note that the fees have been added to the invoice. https://d.pr/i/uVZL1Z
     - Note in WC the parent order doesn't have the fees. https://d.pr/i/X5c08r

**Comment by james-allan** [comment]
I also tested the validating however it seems to have missed updating the shipping tax. 

What I did:

1. Purchased a subscription with shipping. Both had a 10% tax. 
2. Enabled the billing clock.
3. Edited the subscription in Stripe and changed the line item qty and tax data. 
     - As part of this I changed the line item qty and tax. The shipping tax was 10% + a new 5% tax https://d.pr/i/stvUft. 
4. Triggered invoice upcoming.
    - That sent the following repair data.
6. The subscription in Stripe was updated. However, the shipping tax wasn't


```
// This is the product line item repair data.
Array
(
    [price] => price_1JSvi82HTmZdpnU0mIN4xEaX
    [quantity] => 1
    [tax_rates] => Array
        (
            [0] => Array
                (
                    [display_name] => AU-GST-1
                    [inclusive] => 
                    [percentage] => 10
                )

        )

)
// The shipping repair data.
Array
(
    [price_data] => Array
        (
            [currency] => USD
            [product] => prod_KG7lHRcJQMlcSp
            [unit_amount] => 2000
            [recurring] => Array
                (
                    [interval] => month
                    [interval_count] => 1
                )

        )

    [quantity] => 1
)

```

File: includes/subscriptions/class-wc-payments-invoice-service.php [hunk 1]
```diff
@@ -110,136 +166,79 @@ public function maybe_record_first_invoice_payment( int $order_id, string $old_s
...
+					$product              = $item->get_product();
+					$repair_data['price'] = $this->product_service->get_stripe_price_id( $product );
+				} else {
+					$repair_data['price_data'] = WC_Payments_Subscription_Service::format_item_price_data(
+						$subscription->get_currency(),
+						$this->product_service->get_stripe_product_id_for_item( $item->get_type() ),
+						$item->get_total(),
+						$subscription->get_billing_period(),
+						$subscription->get_billing_interval()
+					);
```
Comment (james-allan) [inline includes/subscriptions/class-wc-payments-invoice-service.php:217]: Re this comment: https://github.com/Automattic/woocommerce-payments/pull/2899#issuecomment-922695478

I think we need to add the items tax data similar to how we do it here: https://github.com/Automattic/woocommerce-payments/blob/update/subscription_items/includes/subscriptions/class-wc-payments-subscription-service.php#L637

**Review by james-allan — CHANGES_REQUESTED** [review]

**Comment by chickenn00dle** [comment]
Thanks for looking over this @james-allan!

> Question: Is there a way that we can add something to make this fee distinguishable? I had a look in the Stripe API and I can't see anything to suggest we can add price data meta. So maybe we can't.

We could actually add metadata to the item via [items.metadata](https://stripe.com/docs/api/subscriptions/create#create_subscription-items-metadata).

> This code snippet adds fees to the recurring cart. No fees are charged on the initial order.

Ah. Good catch! I actually wasn't aware of cases where a recurring amount was charged on renewals, but NOT on the initial order 😞 

I'm actually not sure how to deal with this in Stripe. I'm pretty sure any items added via [items](https://stripe.com/docs/api/subscriptions/create#create_subscription-items) will be added to both the initial order AND renewal orders in Stripe.

I guess a follow-up question is, outside of hooks, is it possible to add a recurring amount to a WC subscription that does not appear on the initial order like this?

File: includes/subscriptions/class-wc-payments-subscription-service.php [hunk 2]
```diff
@@ -417,112 +506,258 @@ public function prevent_wcpay_subscription_changes( bool $supported, string $fea
...
+	 * @return array WCPay one time item data.
 	 */
-	private function set_wcpay_subscription_id( WC_Subscription $subscription, string $value ) {
-		$subscription->update_meta_data( self::SUBSCRIPTION_ID_META_KEY, $value );
-		$subscription->save();
+	private function get_one_time_item_data_for_subscription( WC_Subscription $subscription ) : array {
+		$data     = [];
+		$currency = $subscription->get_currency();
+
+		foreach ( $subscription->get_items() as $item ) {
```
[the remark this answers is not in the document]
Comment (chickenn00dle) [reply includes/subscriptions/class-wc-payments-subscription-service.php:656]: Hmm. I don't think this is the case. The `get_one_time_item_data_for_subscription` method is responsible for generating items that should appear ONLY on the first invoice and does not apply to recurring items.

Right now this method checks subscription products associated with a subscription for `sign_up_fee`'s and `one_time_shipping`. As I understand it, it wouldn't matter if we get these products from the subscription or the parent order.

[the remark this answers is not in the document]
Comment (james-allan) [reply includes/subscriptions/class-wc-payments-subscription-service.php:656]: Oh right. So I was working under the assumption that this created the items for the original invoices kinda from scratch but that gap in my understanding was that it's only additive on top of the subscription items. 

That's to say,
Original invoice items = subscription items + these additional signup fees + one time shipping. 

So changing this to the parent order wouldn't have really much effect at all. 🤔

**Comment by james-allan** [comment]
> I guess a follow-up question is, outside of hooks, is it possible to add a recurring amount to a WC subscription that does not appear on the initial order like this?

Free trial and sync'd products which have no initial cost but have recurring prices. (I haven't tested these yet). 

Other than that, it's possible a sign-up fee, or fixed cart coupon is used which reduces the initial invoice cost but not the recurring amount. I'm not sure if that's what you mean though.

File: includes/subscriptions/class-wc-payments-invoice-service.php (hunk 1 above)
Comment (chickenn00dle) [reply includes/subscriptions/class-wc-payments-invoice-service.php:217]: We are actually adding tax data further down in this method. Since this is updating subscription items rather than creating them, we only include the repair data that is needed with the update request.

As for this comment: https://github.com/Automattic/woocommerce-payments/pull/2899#issuecomment-922695478

Good catch again! I believe I've fixed this here: https://github.com/Automattic/woocommerce-payments/pull/2899/commits/afda4e237b2a0408de285c70dd846012badd44b4

Basically, before this change, we were validating tax data by checking that each of our tax rates in WC matched a tax rate in Stripe. But we weren't accounting for cases where Stripe had somehow added an extra tax rate in addition to any that we already were expecting. This change first checks that Stripe and WC have the same number of tax rates per item before iterating through each WC tax rate to compare actual percentages.

**Comment by chickenn00dle** [comment]
> Free trial and sync'd products which have no initial cost but have recurring prices. (I haven't tested these yet).

Great. These cases should be covered since we add a `trial_end` when creating the subscription which will delay the entire first payment altogether 👍 

> Other than that, it's possible a sign-up fee, or fixed cart coupon is used which reduces the initial invoice cost but not the recurring amount.

These cases should also be covered since the recurring item still appears in the initial invoice, but the cost is offset by the coupon discount.

> I'm not sure if that's what you mean though.

Yeah. Sorry. I'm being a bit unclear I think. Essentially, I'm wondering if we can just not support this hook in WCPay subscriptions to avoid unnecessary complexity.

If we do support it, we need to find a way to not have a recurring item appear in the first invoice, but then appear in renewals, which doesn't seem possible with [items](https://stripe.com/docs/api/subscriptions/create#create_subscription-items) without either adding a trial to the subscription as a whole, or adding some discount to the subscription to offset this amount on the first invoice.

**Comment by james-allan** [comment]
> If we do support it, we need to find a way to not have a recurring item appear in the first invoice, but then appear in renewals,

I am not super confident it what options we have and how they work but some ideas might be to:

- Create the subscription without these and then add them to the subscription after the first invoice is created/paid.
- Create the subscription with them but then [delete them from the invoice](https://stripe.com/docs/api/invoiceitems/delete) before we mark it as paid. I'm not sure if the invoice is a draft though so this might not be possible.

**Comment by chickenn00dle** [comment]
Hey @james-allan 

> We could actually add metadata to the item via items.metadata.

I spent some time this afternoon testing this, and while [we can add metadata directly to a subscription item](https://stripe.com/docs/api/subscriptions/create#create_subscription-items-metadata), it doesn't look like this will actually appear in the dashboard anywhere, so this might be pointless, at least assuming you were thinking of adding this for something like helping support.

> Create the subscription with them but then delete them from the invoice before we mark it as paid. I'm not sure if the invoice is a draft though so this might not be possible.

The [initial invoice is created with a status of `open`](https://stripe.com/docs/billing/subscriptions/overview#subscription-lifecycle) which would rule out editing the invoice at this stage.

> Create the subscription without these and then add them to the subscription after the first invoice is created/paid.

This is certainly an option. I am wondering though if the only way we run into these situations is via this hook, would it be better to just not support the hook in V1 and avoid additional work at checkout?

**Comment by chickenn00dle** [comment]
I was hoping to get the discounts portion of this added today, but [I can't seem to get this working for our dev account](pdi1mn-15-p2#comment-35) 😞.

Again, as to not hold anything up, I'm going to set this to needs review and if need be we can address coupons/discounts in a follow up PR.

File: includes/subscriptions/class-wc-payments-product-service.php [hunk 3]
```diff
@@ -113,7 +113,7 @@ public static function get_wcpay_price_hash( WC_Product $product ) : string {
 	 * @param WC_Product $product The product to get the WC Pay price ID for.
 	 * @return string             The product's WC Pay price ID or an empty string.
 	 */
-	public static function get_wcpay_price_id( WC_Product $product ) : string {
+	public function get_wcpay_price_id( WC_Product $product ) : string {
```
Comment (chickenn00dle) [inline includes/subscriptions/class-wc-payments-product-service.php:116]: Just noting that in order for us to mock this method in unit tests, I had to make this non-`static`. This is in large part due to the limitations of subscriptions/subscription products in our unit tests. There might be a better way to handle this, so let me know if you can think of one.

File: tests/unit/subscriptions/test-class-wc-payments-invoice-service.php [hunk 4]
```diff
@@ -324,10 +176,52 @@ public function test_maybe_record_first_invoice_payment() {
...
+				'subscription_item' => 'si_test123',
+				'amount'            => 1000,
+				'quantity'          => 1,
+				'tax_rates'         => [],
+			],
+		];
+
+		$mock_discounts = [];
+
+		$this->mock_product_service
```
Comment (chickenn00dle) [inline tests/unit/subscriptions/test-class-wc-payments-invoice-service.php:202]: See this comment: https://github.com/Automattic/woocommerce-payments/pull/2899/files#r712754395

**Comment by haszari** [comment]
> > Create the subscription without these and then add them to the subscription after the first invoice is created/paid.

> This is certainly an option. I am wondering though if the only way we run into these situations is via this hook, would it be better to just not support the hook in V1 and avoid additional work at checkout?

Good question @chickenn00dle ! @james-allan can you report this as an issue so we can follow up separately. Also if you know more about how this hook is used by merchants - e.g. common use cases - add that info to bug to help us decide how to proceed here.

I'm curious what other similar/related hooks we might need to test and support, any ideas?

**Comment by chickenn00dle** [comment]
Just to expand on this, I did some more digging into [the `woocommerce_cart_calculate_fees` hook](https://docs.woocommerce.com/document/subscriptions/develop/recurring-cart-fees/) and see that by default it does apply fees to the initial order as well, unless the `if ( ! empty( $cart->recurring_cart_key ) )` check is included. 

So we can still support the hook by default, but perhaps note in docs that it's not possible to distinguish between initial and recurring anymore or something like that?

**Comment by james-allan** [comment]
> can you report this as an issue so we can follow up separately.

👍 

> So we can still support the hook by default, but perhaps note in WCPay Subscriptions docs that it's not possible to distinguish between initial and recurring anymore or something like that?

Yeah. The other thing to keep in mind is that this is just an issue in the initial invoice right? The order in WC will be correct, and the customer will be charged the right amount (given we are taking the initial sign up with WC Payments). So it's just a reporting issue in the initial invoice so might not be too big of a deal at all.

File: includes/subscriptions/class-wc-payments-invoice-service.php (hunk 1 above)
Comment (james-allan) [reply includes/subscriptions/class-wc-payments-invoice-service.php:217]: > But we weren't accounting for cases where Stripe had somehow added an extra tax rate in addition to any that we already were expecting.

oooh. Good thing you picked that detail out of my rough steps to replicate. 😅

**Review by james-allan — APPROVED** [review]
Works a treat. Nice work Raz. I just had 1 question on the server PR which if changed will require changes to this PR. But other than that, this looks good.

**Comment by chickenn00dle** [comment]
Thanks @james-allan!

I've renamed these particular product IDs in server as well as in this PR in https://github.com/Automattic/woocommerce-payments/pull/2899/commits/cbd460aeb4427c2dfb70407625b8d89226e07e4d

Note that there are still other instances of `stripe_` in things like getter methods, but since these are mostly unrelated to this PR, I'll go ahead and open a separate issue for renaming these and we can address this later on down the line if that's okay.

**Comment by james-allan** [comment]
> I've renamed these particular product IDs in server as well as in this PR in cbd460a

Awesomesauce 🚀

**Review by james-allan — APPROVED** [review]
